plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1407,85 +1424,93 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
-                  []    -> a
-                  _ : _ -> b
+                case uniListA of
+                    DefaultUniList _ -> pure $ case xs of
+                        []    -> a
+                        _ : _ -> b
+                    _ -> throwing _StructuralUnliftingError "Expected a list but got something else"
             {-# INLINE chooseListDenotation #-}
         in makeBuiltinMeaning
             chooseListDenotation
             (runCostingFunThreeArguments . paramChooseList)
 
     toBuiltinMeaning _semvar MkCons =
         let mkConsDenotation
-                :: SomeConstant uni a -> SomeConstant uni [a] -> EvaluationResult (Opaque val [a])
+                :: SomeConstant uni a -> SomeConstant uni [a] -> BuiltinResult (Opaque val [a])
             mkConsDenotation
               (SomeConstant (Some (ValueOf uniA x)))
               (SomeConstant (Some (ValueOf uniListA xs))) = do
-                  DefaultUniList uniA' <- pure uniListA
-                  -- Checking that the type of the constant is the same as the type of the elements
-                  -- of the unlifted list. Note that there's no way we could enforce this statically
-                  -- since in UPLC one can create an ill-typed program that attempts to prepend
-                  -- a value of the wrong type to a list.
-                  -- Should that rather give us an 'UnliftingError'? For that we need
-                  -- https://github.com/IntersectMBO/plutus/pull/3035
-                  Just Refl <- pure $ uniA `geq` uniA'
-                  pure . fromValueOf uniListA $ x : xs
+                case uniListA of
+                    DefaultUniList uniA' -> case uniA `geq` uniA' of
+                        Just Refl -> pure . fromValueOf uniListA $ x : xs
+                        _         -> throwing _StructuralUnliftingError
+                            "The type of the value does not match the type of elements in the list"
+                    _ -> throwing _StructuralUnliftingError "Expected a list but got something else"
             {-# INLINE mkConsDenotation #-}
         in makeBuiltinMeaning
             mkConsDenotation
             (runCostingFunTwoArguments . paramMkCons)
 
     toBuiltinMeaning _semvar HeadList =
-        let headListDenotation :: SomeConstant uni [a] -> EvaluationResult (Opaque val a)
+        let headListDenotation :: SomeConstant uni [a] -> BuiltinResult (Opaque val a)
             headListDenotation (SomeConstant (Some (ValueOf uniListA xs))) = do
-                DefaultUniList uniA <- pure uniListA
-                x : _ <- pure xs
-                pure $ fromValueOf uniA x
+                case uniListA of
+                    DefaultUniList uniA -> case xs of
+                        []    -> fail "Expected a non-empty list but got an empty one"
+                        x : _ -> pure $ fromValueOf uniA x
+                    _ -> throwing _StructuralUnliftingError "Expected a list but got something else"
```

Comment:
It's awkward that the structural vs operational errors distinction leaks to the user space here, but that's inevitable, since "runtime type errors" are inherently structural and errors like "expected a non-empty list but got an empty one" are inherently operational. 

TODO: this needs documentation.